src/precice/impl/DataContext.cpp

```diff
@@ -126,7 +126,7 @@ int DataContext::mapData(std::optional<double> after, bool skipZero)
           dataDims,
           Eigen::VectorXd::Zero(dataDims * mapping.getOutputMesh()->nVertices())};
 
-      bool skipMapping = skipZero && stample.sample.values.isZero();
+      bool skipMapping = skipZero && (utils::IntraComm::l2norm(stample.sample.values) < math::NUMERICAL_ZERO_DIFFERENCE);
```

Comment:
This is now an unconditional collective for every mapped data, which sounds pretty expensive to me.

An elegant and easy way to prevent this could be:
```suggestion
      bool skipMapping = skipZero && mapping.usesCollectives() && (utils::IntraComm::l2norm(stample.sample.values) < math::NUMERICAL_ZERO_DIFFERENCE);
```

This prevents the `l2norm` collective when `mapping.usesCollectives() == false`.

This could be as simple as
```cpp
// in Mapping.hpp
virtual bool usesCollectives() const = 0;
// for RBF
bool usesCollectives() const override { return true };
// for the NN, NNG and NP (as long as scaled-consistent is locked to serial participants)
bool usesCollectives() const override { return false };
```

Reply:
> This is now an unconditional collective for every mapped data

Actually no, according to the C++ standard, the logical AND (&&) and logical OR (||) operators perform short-circuit evaluation. The boolean is only `true` in the 'initial' skip. In all other cases, the boolean is `false` and the `l2norm` is not evaluated.

Reply:
@fsimonis But your solution no longer checks whether the data is actually zero in case of non-collective mappings?

Reply:
> short-circuit evaluation

:man_facepalming: _true_

So currently, we have a collective for every data-sample when we initialize data.
While not tragic, that still hurts performance for NN/NP etc.

> But your solution no longer checks whether the data is actually zero in case of non-collective mappings?

This should do the trick:

```suggestion
      bool skipMapping = skipZero && ( mapping.usesCollective() ? (utils::IntraComm::l2norm(stample.sample.values) < math::NUMERICAL_ZERO_DIFFERENCE) : stample.sample.values.isZero());
```

Reply:
> While not tragic, that still hurts performance for NN/NP

Let's not worry about this at this point and introduce complexity. It's an initialization cost and for really large cases, establishing the connection and computing the communication maps takes significantly more time (I measured this back then on Supermuc for my mapping experiments). If we observe issues in the future, we can still fix it in the future.

Reply:
Low gain at a low added complexity. I don't have a strong opinion here.
@uekerman your call.

Reply:
> Low gain at a low added complexity. I don't have a strong opinion here.

Well, your logic is incomplete. The PUM doesn't use collectives, whereas the other RBFs do. Also, the code changes might appear trivial, but introduce a dependency of the mapping package to the data context. Modifying one of the mapping to incorporate MPI calls would require changes here as well. If we add GPU to the PUM, we might need such collective MPI calls, how do we handle this then? So we would need separate tests for all cases, which is then yet another todo.

Reply:
I will add a comment to make the short-circuit evaluation more apparent.